pyro/infer/contract.py

```diff
@@ -227,4 +220,15 @@ def contract_to_tensor(tensor_tree, sum_dims, target_ordinal):
     _contract_component(tensor_tree, sum_dims, target_ordinal)
     t, terms = tensor_tree.popitem()
     assert t == target_ordinal
-    return sum(terms)
+    term = sum(terms)
+
+    # Broadcast to any missing iaranges via .expand().
```

Comment:
I've moved broadcasting logic to happen outside of `sumproduct` so as to improve sharing.